### PR TITLE
Build Cleanup

### PR DESCRIPTION
**Pull request by tgvarik**
Update #802 

We have started work on the WASM module build and found it necessary to tidy up the build options so that we can easily disable building things that can’t be built as part of the WASM modules.
- Existing behavior is preserved.
- A few unused preprocessor definitions added with `add_definitions` were removed.
- The `MATERIALX_DISABLE_BUILD_RENDER` option has been deprecated. It will still be honored, but please use `MATERIALX_BUILD_RENDER=OFF` instead.
- Added OPTIONS.md to document most of the options and how some are forced on or off depending on the value of other options.
- The resources and libraries folders are only copied if they are needed.

### CODE REVIEW

**Comment by tgvarik** [comment]
Whoops, I broke `MATERIALX_INSTALL_PYTHON`. Fix shortly.

**Comment by bernardkwok** [comment]
Hi Tom, I noticed sometime during your cleanup modifications that no unit tests are being run anymore: e.g. https://travis-ci.org/autodesk-forks/MaterialX/jobs/659442190

**Comment by tgvarik** [comment]
@bernardkwok I mistook the relevance of the MaterialXTest project. Pushed a change that should fix it. New option to enable building them is on by default now.

**Comment by tgvarik** [comment]
@bernardkwok So the original CMakeLists.txt for MaterialXTest has these lines:
```
# Add language specific generators
if (MATERIALX_BUILD_GEN_ARNOLD)
    LIST(APPEND LIBS "MaterialXGenArnold")
endif()
if (MATERIALX_BUILD_GEN_OGSFX)
    LIST(APPEND LIBS "MaterialXGenOgsFx")
endif()
if (MATERIALX_BUILD_GEN_OGSXML)
    LIST(APPEND LIBS "MaterialXGenOgsXml")
endif()
```
But if ever any of these is not `ON`, building MaterialXTest will fail because the language-specific gen/render tests are not conditionally excluded from MaterialXTest's source. The tests still get built, referencing the unlinked generators. So the linker chokes.

There are really to options:
1) Setting `MATERIALX_BUILD_TEST=ON` forces all generators to build; test files are left as-is, we remove the conditional links.
2) Each generator-specific test file gets wrapped in macro checks to exclude the tests if the generator hasn’t been built.

My thinking is for (2) because it allows for better control over the build, and is what I’d like to push, but I want your input first.

**Comment by tgvarik** [comment]
@bernardkwok I guess there’s a third option: group up the test files by component and conditionally include them in the source. Extra folders under MaterialXTest, each with their own cmakelists.txt (but not necessarily separate build targets). More work but IMO cleaner than (2) without the loss of build control of (1).

**Comment by bernardkwok** [comment]
Hi Tom. We had structure this to run using option 2) originally but this got ugly. It should be option 3) now. I.e. we exclude files from the build inside MaterialXTest in the main cmakefile, but could be we lost this somehow as it's not structure like this. It could also be via subfolder inclusion though may be a bit too much as it's only for 1/2 files for most cases.

**Comment by ashwinbhat** [comment]
Is this source/PyMaterialX/PyMaterialXGenShader/PyUnitConverter.cpp → source/PyMaterialX/PyMaterialXCore/PyUnitConverter.cpp expected?

**Comment by tgvarik** [comment]
Yes, that’s intentional, since the UnitConverter is part of Core now. I didn’t do it completely the first time (forgot to update the `PyModule.cpp`s). Really weird that it still built and tested fine on Mac though (I swear I test before I push!)

Latest push should (hopefully? 🤷‍♂) fix it.

File: CMakeLists.txt [hunk 1]
```diff
@@ -10,81 +10,110 @@ set(CMAKE_POSITION_INDEPENDENT_CODE TRUE)
...
+if(MATERIALX_DISABLE_BUILD_RENDER)
+    message(WARNING "MATERIALX_DISABLE_BUILD_RENDER is deprecated. Please use MATERIALX_BUILD_RENDER instead")
+    set(MATERIALX_BUILD_RENDER OFF)
+endif()
+
+option(MATERIALX_PYTHON_LTO "Enable link-time optimizations for MaterialX Python." ON)
+option(MATERIALX_INSTALL_PYTHON "Install the MaterialX Python package as a third-party library when the install target is built." ON)
 
-# Validation options for compiling and rendering generated shaders
 option(MATERIALX_TEST_RENDER "Run rendering tests for MaterialX Render module. GPU required for graphics validation." OFF)
```
Comment (bernardkwok) [inline CMakeLists.txt]: @tgvarik , as part of the last ILM merge, this should be on all of the time now, but off in CI.

File: source/MaterialXFormat/Util.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+{
+
+string removeExtension(const string& filename)
+{
+  size_t lastDot = filename.find_last_of('.');
+  if (lastDot == string::npos) return filename;
+  return filename.substr(0, lastDot);
+}
+
+bool readFile(const string& filename, string& contents)
```
Comment (bernardkwok) [inline source/MaterialXFormat/Util.cpp]: Small note: Looks like your using different format convention for tabs. We use 4 space tabs.

File: source/MaterialXFormat/Util.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                        const FileSearchPath& searchPath,
+                        DocumentPtr doc,
+                        const StringSet* excludeFiles = nullptr);
+
+/// Load all MaterialX files with given library names in a given path.
+StringVec loadLibraries(const StringVec& libraryNames,
+                        const FilePath& filePath,
+                        DocumentPtr doc,
+                        const StringSet* excludeFiles = nullptr);
+
```
Comment (bernardkwok) [inline source/MaterialXFormat/Util.h]: Misc: remove extra blank lines

File: source/MaterialXTest/CMakeLists.txt [hunk 4]
```diff
@@ -1,41 +1,115 @@
...
+target_link_libraries(MaterialXTest MaterialXCore)
+add_subdirectory(MaterialXFormat)
+target_link_libraries(MaterialXTest MaterialXFormat)
+
+if(MATERIALX_BUILD_GEN)
+  add_subdirectory(MaterialXGenShader)
+  target_link_libraries(MaterialXTest MaterialXGenShader)
+  if(MATERIALX_BUILD_RUNTIME)
+    add_subdirectory(MaterialXRuntime)
+    target_link_libraries(MaterialXTest MaterialXRuntime)
```
Comment (bernardkwok) [inline source/MaterialXTest/CMakeLists.txt]: Just double checking here. We don't need a cmake version bump for this, or any other calls?

File: source/MaterialXTest/MaterialXCore/Units.cpp [hunk 5]
```diff
@@ -153,32 +152,31 @@ TEST_CASE("UnitDocument", "[units]")
...
-                            if (type->isScalar() && type->getBaseType() == mx::TypeDesc::BASETYPE_FLOAT)
+                            if (type == "float")
                             {
                                 float originalval = value->asA<float>();
                                 float convertedValue = uconverter->convert(originalval, input->getUnit(), DISTANCE_DEFAULT);
                                 float reconvert = uconverter->convert(convertedValue, DISTANCE_DEFAULT, input->getUnit());
                                 REQUIRE((originalval - reconvert) < EPSILON);
                             }
-                            else if (type->isFloat2())
+                            else if (type == "vector2" || type == "color2")
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXCore/Units.cpp]: unit conversion is not allowed on color types so the color2, color3, color4 inclusion should be removed.

File: source/MaterialXTest/MaterialXCore/Units.cpp [hunk 6]
```diff
@@ -191,32 +189,32 @@ TEST_CASE("UnitDocument", "[units]")
...
-                            if (type->isScalar() && type->getBaseType() == mx::TypeDesc::BASETYPE_FLOAT)
+                            if (type == "float")
                             {
                                 float originalval = value->asA<float>();
                                 float convertedValue = uconverter->convert(originalval, param->getUnit(), DISTANCE_DEFAULT);
                                 float reconvert = uconverter->convert(convertedValue, DISTANCE_DEFAULT, param->getUnit());
                                 REQUIRE((originalval - reconvert) < EPSILON);
                             }
-                            else if (type->isFloat2())
+                            else if (type == "vector2" || type == "color2")
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXCore/Units.cpp]: Same comment about not allowing color types applies here.

File: source/PyMaterialX/CMakeLists.txt [hunk 7]
```diff
@@ -47,16 +47,22 @@ endif()
...
 add_subdirectory(PyMaterialXFormat)
-add_subdirectory(PyMaterialXGenShader)
-add_subdirectory(PyMaterialXGenGlsl)
-add_subdirectory(PyMaterialXGenOsl)
-if (MATERIALX_BUILD_GEN_OGSFX)
-    add_subdirectory(PyMaterialXGenOgsFx)
-endif()
-if (MATERIALX_BUILD_GEN_ARNOLD)
-    add_subdirectory(PyMaterialXGenArnold)
+if (MATERIALX_BUILD_GEN)
```
Comment (bernardkwok) [inline source/PyMaterialX/CMakeLists.txt]: We'll need to check whether ILM want's this back as originally we had all GLSL and OSL modules as required so would never do this check. I think it should be okay or we we split along "sets" of requirements. e.g. desktop always has GLSL + OSL + MDL. We can figure this out.

File: source/MaterialXGenShader/UnitSystem.h [hunk 8]
```diff
@@ -14,7 +14,7 @@
 #include <MaterialXGenShader/ShaderNode.h>
 #include <MaterialXGenShader/ShaderNodeImpl.h>
 #include <MaterialXGenShader/TypeDesc.h>
-#include <MaterialXGenShader/UnitConverter.h>
+#include <MaterialXCore/UnitConverter.h>
 
 #include <MaterialXCore/Document.h>
```
Comment (bernardkwok) [inline source/MaterialXGenShader/UnitSystem.h]: Can't seem to put a comment there but in the create() method a language is a required input. It's a cached value that only seems to be used for getting the implementation. I think that was one reason why it was in GenShader. I think the easiest thing to do is to not cache this but require it as an input argument to UnitSystem::getImplementationName().

File: source/MaterialXTest/CMakeLists.txt (hunk 4 above)
Comment (tgvarik) [reply source/MaterialXTest/CMakeLists.txt]: No, this should work in 3.1. Multiple calls are allowed, and the linked library target need not be declared in same directory. It definitely works on the build slaves.

File: source/PyMaterialX/CMakeLists.txt (hunk 7 above)
Comment (tgvarik) [reply source/PyMaterialX/CMakeLists.txt]: This is central to our use case -- we need to be able to build without gen at all, and later, to build just with GLSL gen.

File: source/MaterialXGenShader/UnitSystem.h (hunk 8 above)
Comment (tgvarik) [reply source/MaterialXGenShader/UnitSystem.h]: That’s right, the only use is in `UnitSystem::getImplementationName()`. But `getImplementationName` is in turn called by `UnitSystem::supportsTransform()` and `UnitSystem::createNode()`, which would also have to take the language name as argument.

But I’m not sure I see the relevance? We are leaving `UnitSystem` in GenShader; only `UnitConverter` is moving to Core.

File: source/MaterialXTest/MaterialXCore/Units.cpp (hunk 6 above)
Comment (tgvarik) [reply source/MaterialXTest/MaterialXCore/Units.cpp]: Fixed

File: source/MaterialXTest/MaterialXCore/Units.cpp (hunk 5 above)
Comment (tgvarik) [reply source/MaterialXTest/MaterialXCore/Units.cpp]: Fixed

File: source/MaterialXFormat/Util.h (hunk 3 above)
Comment (tgvarik) [reply source/MaterialXFormat/Util.h]: Fixed

File: source/MaterialXFormat/Util.cpp (hunk 2 above)
Comment (tgvarik) [reply source/MaterialXFormat/Util.cpp]: Fixed. 😉

File: CMakeLists.txt (hunk 1 above)
Comment (tgvarik) [reply CMakeLists.txt]: Fixed

File: source/MaterialXGenShader/UnitSystem.h (hunk 8 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/UnitSystem.h]: Yes, your right there is nothing that is required here now.

**Comment by bernardkwok** [comment]
@tgvarik , I tried running this locally and found that the render osl tests do not seem to run (after your change). Can you try this out locally by setting the required osl cmake options ? 
Thanks.

**Comment by bernardkwok** [comment]
@tgvarik , @ashwinbhat : I've merged this in already. If there are additional changes you can add these on top of adsk_contrib/dev.

**Comment by bernardkwok** [comment]
I'm closing this as a version of this has been merged into adsk_contrib/dev and ILM master.
@tgvarik , any modifications or updates feel free to create a new PR.

**Pull request closed without merging**